eCommerce platforms/Adobe Magento/src/code/local/Microsoft/Dfp/controllers/Checkout/OnepageController.php

```diff
@@ -0,0 +1,399 @@
+			$requiredAgreements = Mage::helper('checkout')->getRequiredAgreementIds();
+			if ($requiredAgreements) {
+				$postedAgreements = array_keys($this->getRequest()->getPost('agreement', array()));
+				$diff = array_diff($requiredAgreements, $postedAgreements);
+				if ($diff) {
+					$result['success'] = false;
+					$result['error'] = true;
+					$result['error_messages'] = $this->__('Please agree to all the terms and conditions before placing the order.');
+					$this->getResponse()->setBody(Mage::helper('core')->jsonEncode($result));
+					return;
+				}
+			}
+
+			$data = $this->getRequest()->getPost('payment', array());
+			if ($data) {
+				$data['checks'] = Mage_Payment_Model_Method_Abstract::CHECK_USE_CHECKOUT
+					| Mage_Payment_Model_Method_Abstract::CHECK_USE_FOR_COUNTRY
+					| Mage_Payment_Model_Method_Abstract::CHECK_USE_FOR_CURRENCY
+					| Mage_Payment_Model_Method_Abstract::CHECK_ORDER_TOTAL_MIN_MAX
+					| Mage_Payment_Model_Method_Abstract::CHECK_ZERO_TOTAL;
+				$this->getOnepage()->getQuote()->getPayment()->importData($data);
+			}
+
+			$this->getOnepage()->saveOrder();
+
+			// *********************************************** DFP Changes START *************************************//
+			try {
+				$correlationId = $this->dfp->GUID();
+				$order = Mage::getModel('sales/order')->load($this->getOnepage()->getQuote()->getId(), 'quote_id');
+				$this->dfp->log('************** Invoking Purchase API - START ************* ');
+				$optionArray = Mage::getModel('dfp/config_source_assessmenttypes')->toOptionArray();
+				$assessmentType = $optionArray[Mage::getStoreConfig('dfp/general/assessmenttype')]['label'];
+				$merchantRuleDecision = $this->invokePurchaseDFP($order, $correlationId, $assessmentType);
+				if (
+					strtolower($merchantRuleDecision) == "reject" &&
+					strtolower($assessmentType) == 'protect'
+				) {
+					$this->dfp->log('************** Invoking Purchase API - END ************* ');
+
+					$this->cancelOrder('Order is rejected by DFP.');
+
+					$result['redirect'] = Mage::getUrl('checkout/onepage/failure');
+					$this->getResponse()->setBody(Mage::helper('core')->jsonEncode($result));
+					return;
+				} else {
+					try {
+						$this->dfp->log('************** Invoking Bank Event API - START ************* ');
+						$this->invokeBankEventDFP("Auth", $correlationId);
+						$this->invokeBankEventDFP("Charge", $correlationId);
+						$this->dfp->log('************** Invoking Bank Event API - END ************* ');
+					} catch (exception $e) {
+						$this->dfp->log($e->getMessage());
+					}
+
+					$redirectUrl = $this->getOnepage()->getCheckout()->getRedirectUrl();
+					$result['success'] = true;
+					$result['error']   = false;
+				}
+				$this->dfp->log('************** Invoking Purchase API - END ************* ');
+			} catch (exception $e) {
+				$this->dfp->log($e->getMessage());
+				$redirectUrl = $this->getOnepage()->getCheckout()->getRedirectUrl();
+				$result['success'] = true;
```

Comment:
success = true and error = false in an catch? That doesn't seem quite right to me..

Reply:
So thanks for asking.  It's on purpose to not have DFP failing also break Magento checkout.  e.g. a business decision.  I suppose this could later on be part of configuration so that each merchant could choose what happens when DFP isn't working.  I imagine they would prefer this behavior though and would simply do more manual review if they had to rather than flat out lose/miss sales.